# Task 4.9 Intro to Data Visualization with Python / PART 1

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Creating a path
path = r'/Users/ColinLynt/11-2021 Instacart Basket Analysis'

In [3]:
# Importing orders_products_revised_2
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_revised_2.pkl'))

In [4]:
# Importing customer data set
customer_db = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

### Question 4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [5]:
# Checking rows and columns
customer_db.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
# Checking data for clarity
customer_db.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [7]:
# Changing the names of columns
customer_db.rename(columns = {'Surnam':'Last Name','STATE':'State','date_joined':'Date Joined','fam_status':'Marital Status','income':'Income'}, inplace = True)

In [8]:
# Checking rows and columns after rename
customer_db.head()

user_id First Name Last Name  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants Marital Status  Income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

### Question 5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [9]:
# Checking data for clarity
customer_db.describe()

user_id            Age   n_dependants         Income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [10]:
#Checking the number of rows and columns within df_prods
customer_db.shape

(206209, 10)

In [11]:
#Cheack for mixed types customer_db

for col in customer_db.columns.tolist():
  weird = (customer_db[[col]].applymap(type) != customer_db[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer_db[weird]) > 0:
    print (col)

First Name


In [12]:
#Checking for null/missing values in customer_db
customer_db.isnull().sum()

user_id               0
First Name        11259
Last Name             0
Gender                0
State                 0
Age                   0
Date Joined           0
n_dependants          0
Marital Status        0
Income                0
dtype: int64

#### There are 11,259 first names missing from the database.

In [13]:
# Dropping First Name from customer_db
customer_db = customer_db.drop(columns = ['First Name'])

#### Droped First Name coumn from database as it's not necessary and has to many missing values.

In [14]:
# Creating duplicate data db
customer_db_dup = customer_db[customer_db.duplicated()]

In [15]:
# Checking for duplicate data
customer_db_dup.shape

(0, 9)

#### There is no duplicate data within the customers_db

### Question 6. Combine your customer data with the rest of your prepared Instacart data. Tip: Make sure the key columns are of the same data type!

In [16]:
# Checking the names of rows of customer_db
customer_db.head(5)

user_id Last Name  Gender       State  Age Date Joined  n_dependants  \
0    26711  Esquivel  Female    Missouri   48    1/1/2017             3   
1    33890      Hart  Female  New Mexico   36    1/1/2017             0   
2    65803    Farley    Male       Idaho   35    1/1/2017             2   
3   125935     Hicks  Female        Iowa   40    1/1/2017             0   
4   130797   Gilmore  Female    Maryland   26    1/1/2017             1   

  Marital Status  Income  
0        married  165665  
1         single   59285  
2        married   99568  
3         single   42049  
4        married   40374

In [19]:
# Checking the names of rows in ords_prods_merge
ords_prods_merge.head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda      77.0            7.0     9.0   
1  Organic Unsweetened Vanilla Almond Milk      91.0           16.0    12.5   
2                      Original Beef Jerky      23.0           19.0     4.4   
3               Aged White Cheddar Popcorn      23.0           19.0     4.7   
4         XL Pick-A-Size Paper Towel Rolls      54.0           17.0     1.0   

  _merge  max_order  loyalty_flag  spending spending_flag  frequency  \
0   both         10  New customer  6.367797   Low Spender       20.5   
1   both         10  New customer  6.367797   Low Spender       20.5   
2   both         10  New customer  6.367797   Low Spender       20.5   
3   both         10  New customer  6.367797   Low Spender       20.5   
4   both         10  New customer  6.367797   Low Spender       20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer

In [20]:
# Dropping _merge from the ords_prods_merge db to ensure the merge can happen
ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [22]:
ords_prods_merge.shape

(32436241, 19)

In [23]:
customer_db.shape

(206209, 9)

In [24]:
#Merging customer_db and ords_prods_merge
db_merge = ords_prods_merge.merge(customer_db, on = ['user_id'], indicator = True)

In [25]:
# Checking the merge
db_merge.head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...         frequency_flag  \
0                                     Soda  ...  Non-frequent customer   
1  Organic Unsweetened Vanilla Almond Milk  ...  Non-frequent customer   
2                      Original Beef Jerky  ...  Non-frequent customer   
3               Aged White Cheddar Popcorn  ...  Non-frequent customer   
4         XL Pick-A-Size Paper Towel Rolls  ...  Non-frequent customer   

   Last Name  Gender    State Age  Date Joined n_dependants  Marital Status  \
0     Nguyen  Female  Alabama  31    2/17/2019            3         married   
1     Nguyen  Female  Alabama  31    2/17/2019            3         married   
2     Nguyen  Female  Alabama  31    2/17/2019            3         married   
3     Nguyen  Female  Alabama  31    2/17/2019            3         married   
4     Nguyen  Female  Alabama  31    2/17/2019            3         married   

  Income _merge  
0  40423   both  
1  40423   both  
2  40423   both  
3  40423   both  
4  40423   both  

[5 rows x 28 columns]

In [26]:
# Export data to pkl
db_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data','db_merge.pkl'))